dialogue-clients/src/test/java/com/palantir/dialogue/clients/DialogueClientsDnsResolutionTest.java

```diff
@@ -0,0 +1,180 @@
+import javax.net.ssl.X509TrustManager;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class DialogueClientsDnsResolutionTest {
+    private static final String FOO_PATH = "/foo1";
+    private Undertow undertow;
+    private HttpHandler undertowHandler;
+
+    @BeforeEach
+    public void before() {
+        SSLContext sslContext = SslSocketFactories.createSslContext(TestConfigurations.SSL_CONFIG);
+        undertow = Undertow.builder()
+                .addHttpsListener(
+                        0,
+                        "localhost",
+                        sslContext,
+                        new BlockingHandler(exchange -> undertowHandler.handleRequest(exchange)))
+                .build();
+        undertow.start();
+    }
+
+    @AfterEach
+    public void after() {
+        if (undertow != null) {
+            undertow.stop();
+        }
+    }
+
+    @Test
+    void reloadingFactoryDnsTaskCleanup() {
+        List<String> requestPaths = Collections.synchronizedList(new ArrayList<>());
+        undertowHandler = exchange -> {
+            requestPaths.add(exchange.getRequestPath());
+            exchange.setStatusCode(200);
+        };
+
+        TaggedMetricRegistry registry = new DefaultTaggedMetricRegistry();
+        Refreshable<ServicesConfigBlock> refreshable = Refreshable.only(ServicesConfigBlock.builder()
+                .defaultSecurity(TestConfigurations.SSL_CONFIG)
+                .putServices(
+                        "foo",
+                        PartialServiceConfiguration.builder()
+                                .addUris(getUri(undertow) + FOO_PATH)
+                                .build())
+                .build());
+        DialogueClients.ReloadingFactory factory =
+                DialogueClients.create(refreshable).withUserAgent(TestConfigurations.AGENT);
+
+        @SuppressWarnings("unused")
+        SampleServiceBlocking client = factory.get(SampleServiceBlocking.class, "foo");
+        WeakReference<SampleServiceBlocking> clientWeakRef = new WeakReference<>(client);
+        client.voidToVoid();
+        assertThat(requestPaths).containsExactly("/foo1/voidToVoid");
+
+        Counter activeDnsRefreshTasks =
+                DialogueDnsDiscoveryMetrics.of(registry).scheduled("dialogue-nonreloading-SampleServiceBlocking");
+        assertThat(activeDnsRefreshTasks.getCount()).isEqualTo(1L);
+        assertThat(clientWeakRef.get())
+                .as("This reference is expected to be present before the strong client ref is unset")
+                .isNotNull();
+        // This reference _must_ be freed to allow the cleaner to do its job.
+        client = null;
+        attemptToGarbageCollect(clientWeakRef);
+        Awaitility.waitAtMost(Duration.ofSeconds(15)).untilAsserted(() -> assertThat(activeDnsRefreshTasks.getCount())
+                .isEqualTo(0L));
+    }
```

Comment:
This test is expected to be failing at this point, I haven't implemented dns refreshing on this codepath quite yet